File: go/clients/streaming.go
package clients

import (
	"bufio"
	"bytes"
	"encoding/json"
	"io"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/DexwoxBusiness/dexcost-go/core"
	"github.com/DexwoxBusiness/dexcost-go/pricing"
	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

// StreamProvider is the LLM provider behind a streaming response. The wire
// format is the same shape per provider as the unary helpers — see
// RecordOpenAIResponse / RecordAnthropicResponse for the (non-streaming)
// equivalents.
type StreamProvider int

const (
	// StreamProviderOpenAI parses OpenAI-style SSE chunks: `data: {...}\n\n`
	// where each JSON payload is a `chat.completion.chunk` and the final
	// chunk carries `usage.{prompt_tokens,completion_tokens}` when the
	// caller passes `stream_options: {include_usage: true}`.
	StreamProviderOpenAI StreamProvider = iota
	// StreamProviderAnthropic parses Anthropic-style SSE chunks. Anthropic
	// emits two relevant event types: `message_start` (carries `model` +
	// initial `usage.input_tokens`) and `message_delta` (carries cumulative
	// `usage.output_tokens` up to the latest delta).
	StreamProviderAnthropic
	// StreamProviderGroq is OpenAI-compatible; identical wire format,
	// stamped with `provider = "groq"` on the recorded event.
	StreamProviderGroq
)

// StreamRecorder wraps an SSE response body to accumulate token usage from
// streaming chunks and record an llm_call event when the stream is fully
// consumed (or closed by the caller). It implements `io.ReadCloser` so it can
// be substituted in place of the original `*http.Response.Body`.
//
// The wrapper is transparent — `Read` returns bytes byte-for-byte from the
// underlying stream — but tees buffered reads through an SSE chunk parser.
// When the underlying stream returns `io.EOF` (or the wrapper is `Close`d
// before EOF), the accumulated usage is recorded as a single llm_call event
// against the supplied `taskID`. Subsequent `Close` calls are no-ops.
//
// This mirrors the Python streaming pattern at `instruments/openai.py:257`
// (sync) and `:313` (async) — wrap the iterator, watch each chunk for
// `model`+`usage`, finalize on stream end.
type StreamRecorder struct {
	body         io.ReadCloser
	provider     StreamProvider
	buffer       core.Buffer
	pricingEng   *pricing.Engine
	taskID       uuid.UUID
	requestModel string

	startTime time.Time

	scanBuf bytes.Buffer
	model   string
	usage   map[string]int

	mu        sync.Mutex
	finalized bool
}

// NewOpenAIStreamRecorder wraps an OpenAI streaming chat completion response.
// `requestModel` is used as a fallback when chunks don't carry a `model`
// field. Pass the body verbatim from `*http.Response.Body`; the recorder
// implements `io.ReadCloser` and forwards reads transparently.
func NewOpenAIStreamRecorder(
	body io.ReadCloser,
	bufferStore core.Buffer,
	pricingEng *pricing.Engine,
	taskID uuid.UUID,
	requestModel string,
) *StreamRecorder {
	return newStreamRecorder(body, StreamProviderOpenAI, bufferStore, pricingEng, taskID, requestModel)
}

// NewAnthropicStreamRecorder wraps an Anthropic streaming messages response.
func NewAnthropicStreamRecorder(
	body io.ReadCloser,
	bufferStore core.Buffer,
	pricingEng *pricing.Engine,
	taskID uuid.UUID,
	requestModel string,
) *StreamRecorder {
	return newStreamRecorder(body, StreamProviderAnthropic, bufferStore, pricingEng, taskID, requestModel)
}

// NewGroqStreamRecorder wraps a Groq streaming chat completion response.
// Groq is wire-compatible with OpenAI; this constructor exists so the
// recorded event carries `provider = "groq"`.
func NewGroqStreamRecorder(
	body io.ReadCloser,
	bufferStore core.Buffer,
	pricingEng *pricing.Engine,
	taskID uuid.UUID,
	requestModel string,
) *StreamRecorder {
	return newStreamRecorder(body, StreamProviderGroq, bufferStore, pricingEng, taskID, requestModel)
}

func newStreamRecorder(
	body io.ReadCloser,
	provider StreamProvider,
	bufferStore core.Buffer,
	pricingEng *pricing.Engine,
	taskID uuid.UUID,
	requestModel string,
) *StreamRecorder {
	return &StreamRecorder{
		body:         body,
		provider:     provider,
		buffer:       bufferStore,
		pricingEng:   pricingEng,
		taskID:       taskID,
		requestModel: requestModel,
		startTime:    time.Now(),
		usage:        make(map[string]int),
	}
}

// Read forwards bytes from the underlying body and tees them through the
// SSE chunk parser. On any terminal error (EOF or otherwise — e.g. the
// server hangs up mid-stream with io.ErrUnexpectedEOF) the accumulated
// usage is recorded so callers that drain via io.ReadAll without an
// explicit Close don't silently lose the event.
func (s *StreamRecorder) Read(p []byte) (int, error) {
	n, err := s.body.Read(p)
	if n > 0 {
		s.scanBuf.Write(p[:n])
		s.consumeBuffered()
	}
	if err != nil {
		s.finalize()
	}
	return n, err
}

// Close ends the stream early and finalizes the event. Safe to call multiple
// times — subsequent calls are no-ops.
func (s *StreamRecorder) Close() error {
	s.finalize()
	return s.body.Close()
}

// Usage returns the accumulated input/output/cached token counts. Useful for
// tests and for callers who want to inspect the recorder after consumption.
func (s *StreamRecorder) Usage() (input, output, cached int) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.usage["input"], s.usage["output"], s.usage["cached"]
}

// Model returns the model name observed during streaming. If the stream
// never emitted a model field, the fallback `requestModel` is returned.
func (s *StreamRecorder) Model() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.model != "" {
		return s.model
	}
	return s.requestModel
}

// consumeBuffered scans the internal buffer for complete SSE events
// (terminated by a blank line) and feeds each one through the chunk parser.
// Partial events at the tail are kept in the buffer for the next call.
func (s *StreamRecorder) consumeBuffered() {
	data := s.scanBuf.Bytes()
	for {
		idx := bytes.Index(data, []byte("\n\n"))
		if idx < 0 {
			break
		}
		event := data[:idx]
		data = data[idx+2:]
		s.processSSEEvent(event)
	}
	// Re-buffer the unconsumed tail.
	s.scanBuf.Reset()
	s.scanBuf.Write(data)
}

// processSSEEvent extracts the JSON `data:` lines from an SSE event block
// and dispatches them to the provider-specific chunk processor.
func (s *StreamRecorder) processSSEEvent(event []byte) {
	scanner := bufio.NewScanner(bytes.NewReader(event))
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if !strings.HasPrefix(line, "data:") {
			continue
		}
		payload := strings.TrimSpace(strings.TrimPrefix(line, "data:"))
		if payload == "" || payload == "[DONE]" {
			continue
		}
		var chunk map[string]interface{}
		if err := json.Unmarshal([]byte(payload), &chunk); err != nil {
			continue
		}
		switch s.provider {
		case StreamProviderOpenAI, StreamProviderGroq:
			s.processOpenAIChunk(chunk)
		case StreamProviderAnthropic:
			s.processAnthropicChunk(chunk)
		}
	}
}

// processOpenAIChunk pulls model + usage from an OpenAI-shaped chunk.
// Usage typically appears only on the final chunk when the caller passes
// `stream_options: {include_usage: true}`.
func (s *StreamRecorder) processOpenAIChunk(chunk map[string]interface{}) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if m, ok := chunk["model"].(string); ok && m != "" {
		s.model = m
	}
	if usage, ok := chunk["usage"].(map[string]interface{}); ok {
		if v := intFromMap(usage, "prompt_tokens"); v > 0 {
			s.usage["input"] = v
		}
		if v := intFromMap(usage, "completion_tokens"); v > 0 {
			s.usage["output"] = v
		}
		if v := intFromMap(usage, "cached_tokens"); v > 0 {
			s.usage["cached"] = v
		}
	}
}

// processAnthropicChunk handles Anthropic's typed event stream. The
// `message_start` event carries `message.model` and the initial
// `message.usage.input_tokens`; subsequent `message_delta` events emit
// cumulative `usage.output_tokens` up to the latest delta.
func (s *StreamRecorder) processAnthropicChunk(chunk map[string]interface{}) {
	s.mu.Lock()
	defer s.mu.Unlock()
	eventType, _ := chunk["type"].(string)
	switch eventType {
	case "message_start":
		message, ok := chunk["message"].(map[string]interface{})
		if !ok {
			return
		}
		if m, ok := message["model"].(string); ok && m != "" {
			s.model = m
		}
		if usage, ok := message["usage"].(map[string]interface{}); ok {
			if v := intFromMap(usage, "input_tokens"); v > 0 {
				s.usage["input"] = v
			}
			if v := intFromMap(usage, "cache_read_input_tokens"); v > 0 {
				s.usage["cached"] = v
			}
		}
	case "message_delta":
		if usage, ok := chunk["usage"].(map[string]interface{}); ok {
			if v := intFromMap(usage, "output_tokens"); v > 0 {
				s.usage["output"] = v
			}
			if v := intFromMap(usage, "input_tokens"); v > 0 && s.usage["input"] == 0 {
				s.usage["input"] = v
			}
		}
	}
}

// finalize records the accumulated usage as a single llm_call event. Idempotent.
func (s *StreamRecorder) finalize() {
	s.mu.Lock()
	if s.finalized {
		s.mu.Unlock()
		return
	}
	s.finalized = true
	model := s.model
	if model == "" {
		model = s.requestModel
	}
	input := s.usage["input"]
	output := s.usage["output"]
	cached := s.usage["cached"]
	s.mu.Unlock()

	latencyMs := int(time.Since(s.startTime).Milliseconds())

	event := core.NewEvent(s.taskID, core.EventTypeLLMCall)
	event.Provider = s.providerName()
	event.Model = model
	event.LatencyMs = &latencyMs

	if input == 0 && output == 0 {
		event.CostUSD = decimal.Zero
		event.CostConfidence = core.CostConfidenceUnknown
		event.PricingSource = core.PricingSourceUnknown
		event.Details["streaming"] = true
	} else {
		costResult := s.pricingEng.GetCost(model, input, output, cached, 0)
		event.CostUSD = costResult.CostUSD
		event.CostConfidence = core.CostConfidence(costResult.CostConfidence)
		event.PricingSource = core.PricingSource(costResult.PricingSource)
		event.PricingVersion = costResult.PricingVersion
		event.InputTokens = intPtr(input)
		event.OutputTokens = intPtr(output)
		if cached > 0 {
			event.CachedTokens = intPtr(cached)
		}
		event.Details["streaming"] = true
	}

	if err := s.buffer.InsertEvent(event); err != nil {
		// finalize runs in the read path, so surfacing this error would force
		// callers to swallow it. Mirror the Python helper's debug-log behavior.
		log.Printf("[dexcost] streaming finalize: insert event failed: %v", err)
	}
}

func (s *StreamRecorder) providerName() string {
	switch s.provider {
	case StreamProviderOpenAI:
		return "openai"
	case StreamProviderAnthropic:
		return "anthropic"
	case StreamProviderGroq:
		return "groq"
	}
	return "unknown"
}

File: rust/src/clients/streaming.rs
//! SSE streaming aggregators for OpenAI and Anthropic chat completions.
//!
//! Mirrors the Python `dexcost.clients.tracked_openai_stream` and
//! `tracked_anthropic_stream` helpers. The aggregators consume a stream of
//! Server-Sent Events (`data: ...` lines) and incrementally accumulate token
//! usage. When the stream ends, callers can produce a synthetic response map
//! suitable for [`crate::clients::wrappers::record_openai_response`] /
//! [`crate::clients::wrappers::record_anthropic_response`].
//!
//! # Feature flag
//!
//! This module is gated behind the `streaming` feature.
//!
//! # Typical usage
//!
//! ```ignore
//! use bytes::Bytes;
//! use futures::StreamExt;
//! use dexcost::clients::streaming::{drain_openai_stream};
//! use dexcost::clients::wrappers::record_openai_response;
//!
//! # async fn _example(
//! #     mut buffer: dexcost::transport::buffer::EventBuffer,
//! #     pricing: dexcost::pricing::engine::PricingEngine,
//! #     stream: impl futures::Stream<Item = Result<Bytes, std::io::Error>> + Unpin,
//! # ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
//! let synth = drain_openai_stream(stream).await?;
//! record_openai_response(&mut buffer, &pricing, "task-123", &synth).await?;
//! # Ok(())
//! # }
//! ```

use std::pin::Pin;

use bytes::Bytes;
use futures::Stream;
use futures::StreamExt;
use serde_json::{json, Value};

/// Error returned while aggregating an SSE stream.
#[derive(Debug, thiserror::Error)]
pub enum StreamAggregateError {
    #[error("stream transport error: {0}")]
    Transport(String),
    #[error("invalid UTF-8 in SSE chunk: {0}")]
    Utf8(#[from] std::str::Utf8Error),
}

/// In-memory aggregator for OpenAI chat-completion SSE streams.
///
/// OpenAI emits one `data: {...}` event per chunk. The terminal `data: [DONE]`
/// event is treated as end-of-stream. Token usage is reported on the final
/// chunk when the request was made with `stream_options.include_usage = true`.
#[derive(Default, Debug, Clone)]
pub struct OpenAIStreamAggregator {
    pub model: Option<String>,
    pub prompt_tokens: Option<i64>,
    pub completion_tokens: Option<i64>,
    pub cached_tokens: Option<i64>,
    pub finished: bool,
}

impl OpenAIStreamAggregator {
    pub fn new() -> Self {
        Self::default()
    }

    /// Feed one JSON event extracted from a `data: ...` SSE line.
    pub fn feed(&mut self, value: &Value) {
        if self.model.is_none() {
            if let Some(m) = value.get("model").and_then(|v| v.as_str()) {
                self.model = Some(m.to_string());
            }
        }

        if let Some(usage) = value.get("usage") {
            if let Some(n) = usage.get("prompt_tokens").and_then(|v| v.as_i64()) {
                self.prompt_tokens = Some(n);
            }
            if let Some(n) = usage.get("completion_tokens").and_then(|v| v.as_i64()) {
                self.completion_tokens = Some(n);
            }
            if let Some(c) = usage
                .get("prompt_tokens_details")
                .and_then(|d| d.get("cached_tokens"))
                .and_then(|v| v.as_i64())
            {
                self.cached_tokens = Some(c);
            }
        }
    }

    /// Return a synthetic non-streaming response shape suitable for
    /// [`crate::clients::wrappers::record_openai_response`].
    pub fn into_response_map(self) -> Value {
        let model = self.model.unwrap_or_else(|| "unknown".to_string());
        let mut usage = json!({
            "prompt_tokens": self.prompt_tokens.unwrap_or(0),
            "completion_tokens": self.completion_tokens.unwrap_or(0),
        });
        if let Some(cached) = self.cached_tokens {
            usage["prompt_tokens_details"] = json!({ "cached_tokens": cached });
        }
        json!({ "model": model, "usage": usage })
    }
}

/// In-memory aggregator for Anthropic message SSE streams.
///
/// The relevant events are:
/// - `message_start` — carries the initial `usage.input_tokens` and `model`.
/// - `message_delta` — the *final* `usage.output_tokens` (cumulative).
/// - `message_stop` — terminator.
#[derive(Default, Debug, Clone)]
pub struct AnthropicStreamAggregator {
    pub model: Option<String>,
    pub input_tokens: Option<i64>,
    pub output_tokens: Option<i64>,
    pub cache_read_input_tokens: Option<i64>,
    pub finished: bool,
}

impl AnthropicStreamAggregator {
    pub fn new() -> Self {
        Self::default()
    }

    /// Feed one JSON event extracted from a `data: ...` SSE line.
    pub fn feed(&mut self, value: &Value) {
        let event_type = value.get("type").and_then(|v| v.as_str()).unwrap_or("");
        match event_type {
            "message_start" => {
                if let Some(message) = value.get("message") {
                    if self.model.is_none() {
                        if let Some(m) = message.get("model").and_then(|v| v.as_str()) {
                            self.model = Some(m.to_string());
                        }
                    }
                    if let Some(usage) = message.get("usage") {
                        if let Some(n) = usage.get("input_tokens").and_then(|v| v.as_i64()) {
                            self.input_tokens = Some(n);
                        }
                        if let Some(n) = usage.get("output_tokens").and_then(|v| v.as_i64()) {
                            self.output_tokens = Some(n);
                        }
                        if let Some(n) = usage
                            .get("cache_read_input_tokens")
                            .and_then(|v| v.as_i64())
                        {
                            self.cache_read_input_tokens = Some(n);
                        }
                    }
                }
            }
            "message_delta" => {
                if let Some(usage) = value.get("usage") {
                    if let Some(n) = usage.get("output_tokens").and_then(|v| v.as_i64()) {
                        // The delta carries the *final* cumulative output count.
                        self.output_tokens = Some(n);
                    }
                }
            }
            "message_stop" => {
                self.finished = true;
            }
            _ => {}
        }
    }

    /// Return a synthetic non-streaming response shape suitable for
    /// [`crate::clients::wrappers::record_anthropic_response`].
    pub fn into_response_map(self) -> Value {
        let model = self.model.unwrap_or_else(|| "unknown".to_string());
        let mut usage = json!({
            "input_tokens": self.input_tokens.unwrap_or(0),
            "output_tokens": self.output_tokens.unwrap_or(0),
        });
        if let Some(cached) = self.cache_read_input_tokens {
            usage["cache_read_input_tokens"] = json!(cached);
        }
        json!({ "model": model, "usage": usage })
    }
}

// ---------------------------------------------------------------------------
// SSE line parsing helpers
// ---------------------------------------------------------------------------

/// Iterate over the `data:` payloads in a buffered SSE chunk and pass each as
/// a JSON value to `feed`. Skips the SSE terminator `[DONE]`.
fn feed_sse_chunk<F>(buffer: &str, mut feed: F)
where
    F: FnMut(&Value),
{
    for line in buffer.lines() {
        // RFC 8895: lines starting with ":" are comments; ignore.
        if line.starts_with(':') {
            continue;
        }
        let payload = match line.strip_prefix("data:") {
            Some(p) => p.trim_start(),
            None => continue,
        };
        if payload.is_empty() || payload == "[DONE]" {
            continue;
        }
        if let Ok(value) = serde_json::from_str::<Value>(payload) {
            feed(&value);
        }
    }
}

/// Drain an OpenAI SSE byte stream, returning a synthetic response map ready
/// to be passed to [`crate::clients::wrappers::record_openai_response`].
///
/// Any transport error short-circuits the stream and is propagated.
pub async fn drain_openai_stream<S, E>(
    mut stream: Pin<Box<S>>,
) -> Result<Value, StreamAggregateError>
where
    S: Stream<Item = Result<Bytes, E>> + ?Sized,
    E: std::fmt::Display,
{
    let mut agg = OpenAIStreamAggregator::new();
    let mut text_buf = String::new();
    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| StreamAggregateError::Transport(e.to_string()))?;
        let s = std::str::from_utf8(&chunk)?;
        text_buf.push_str(s);

        // Process every complete event (terminated by a blank line).
        while let Some(idx) = text_buf.find("\n\n") {
            let event = text_buf[..idx].to_string();
            text_buf.drain(..idx + 2);
            feed_sse_chunk(&event, |v| agg.feed(v));
        }
    }
    if !text_buf.is_empty() {
        feed_sse_chunk(&text_buf, |v| agg.feed(v));
    }
    Ok(agg.into_response_map())
}

/// Drain an Anthropic SSE byte stream, returning a synthetic response map
/// ready to be passed to
/// [`crate::clients::wrappers::record_anthropic_response`].
pub async fn drain_anthropic_stream<S, E>(
    mut stream: Pin<Box<S>>,
) -> Result<Value, StreamAggregateError>
where
    S: Stream<Item = Result<Bytes, E>> + ?Sized,
    E: std::fmt::Display,
{
    let mut agg = AnthropicStreamAggregator::new();
    let mut text_buf = String::new();
    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| StreamAggregateError::Transport(e.to_string()))?;
        let s = std::str::from_utf8(&chunk)?;
        text_buf.push_str(s);
        while let Some(idx) = text_buf.find("\n\n") {
            let event = text_buf[..idx].to_string();
            text_buf.drain(..idx + 2);
            feed_sse_chunk(&event, |v| agg.feed(v));
        }
    }
    if !text_buf.is_empty() {
        feed_sse_chunk(&text_buf, |v| agg.feed(v));
    }
    Ok(agg.into_response_map())
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use futures::stream;
    use serde_json::json;

    #[test]
    fn openai_aggregator_collects_model_and_usage() {
        let mut agg = OpenAIStreamAggregator::new();
        agg.feed(&json!({
            "id": "chatcmpl-1",
            "model": "gpt-4o",
            "choices": [{ "delta": { "content": "Hello" } }],
        }));
        agg.feed(&json!({
            "id": "chatcmpl-1",
            "model": "gpt-4o",
            "choices": [{ "delta": { "content": " world" } }],
        }));
        agg.feed(&json!({
            "id": "chatcmpl-1",
            "model": "gpt-4o",
            "usage": {
                "prompt_tokens": 10,
                "completion_tokens": 5,
                "prompt_tokens_details": { "cached_tokens": 2 }
            }
        }));

        let synth = agg.into_response_map();
        assert_eq!(synth["model"], "gpt-4o");
        assert_eq!(synth["usage"]["prompt_tokens"], 10);
        assert_eq!(synth["usage"]["completion_tokens"], 5);
        assert_eq!(synth["usage"]["prompt_tokens_details"]["cached_tokens"], 2);
    }

    #[test]
    fn openai_aggregator_handles_no_usage() {
        let mut agg = OpenAIStreamAggregator::new();
        agg.feed(&json!({
            "model": "gpt-4o-mini",
            "choices": [{ "delta": { "content": "hi" } }]
        }));
        let synth = agg.into_response_map();
        assert_eq!(synth["model"], "gpt-4o-mini");
        assert_eq!(synth["usage"]["prompt_tokens"], 0);
        assert_eq!(synth["usage"]["completion_tokens"], 0);
        assert!(synth["usage"].get("prompt_tokens_details").is_none());
    }

    #[test]
    fn anthropic_aggregator_collects_input_and_output_from_events() {
        let mut agg = AnthropicStreamAggregator::new();
        agg.feed(&json!({
            "type": "message_start",
            "message": {
                "id": "msg_1",
                "model": "claude-3-5-sonnet-20241022",
                "usage": {
                    "input_tokens": 12,
                    "output_tokens": 1,
                    "cache_read_input_tokens": 4
                }
            }
        }));
        agg.feed(&json!({
            "type": "content_block_delta",
            "index": 0,
            "delta": { "type": "text_delta", "text": "Hello" }
        }));
        agg.feed(&json!({
            "type": "message_delta",
            "delta": { "stop_reason": "end_turn" },
            "usage": { "output_tokens": 27 }
        }));
        agg.feed(&json!({ "type": "message_stop" }));

        assert!(agg.finished);
        let synth = agg.into_response_map();
        assert_eq!(synth["model"], "claude-3-5-sonnet-20241022");
        assert_eq!(synth["usage"]["input_tokens"], 12);
        // Final cumulative count, not the placeholder from message_start.
        assert_eq!(synth["usage"]["output_tokens"], 27);
        assert_eq!(synth["usage"]["cache_read_input_tokens"], 4);
    }

    #[test]
    fn anthropic_aggregator_handles_missing_message_start() {
        // Some integrators only forward delta events; aggregator must not panic.
        let mut agg = AnthropicStreamAggregator::new();
        agg.feed(&json!({
            "type": "message_delta",
            "delta": { "stop_reason": "end_turn" },
            "usage": { "output_tokens": 4 }
        }));
        let synth = agg.into_response_map();
        assert_eq!(synth["model"], "unknown");
        assert_eq!(synth["usage"]["input_tokens"], 0);
        assert_eq!(synth["usage"]["output_tokens"], 4);
    }

    #[test]
    fn feed_sse_chunk_skips_done_and_comments() {
        let mut events: Vec<Value> = Vec::new();
        let chunk = ": keep-alive\n\
             data: {\"x\":1}\n\
             \n\
             data: {\"x\":2}\n\
             data: [DONE]\n";
        feed_sse_chunk(chunk, |v| events.push(v.clone()));
        assert_eq!(events.len(), 2);
        assert_eq!(events[0], json!({ "x": 1 }));
        assert_eq!(events[1], json!({ "x": 2 }));
    }

    #[tokio::test]
    async fn drain_openai_stream_aggregates_usage() {
        // Three SSE chunks: two delta events and one usage event, each followed
        // by the SSE event delimiter (\n\n).
        let chunks: Vec<Result<Bytes, std::io::Error>> = vec![
            Ok(Bytes::from(
                "data: {\"model\":\"gpt-4o\",\"choices\":[{\"delta\":{\"content\":\"Hi\"}}]}\n\n",
            )),
            Ok(Bytes::from(
                "data: {\"model\":\"gpt-4o\",\"choices\":[{\"delta\":{\"content\":\" \"}}]}\n\n",
            )),
            Ok(Bytes::from(
                "data: {\"model\":\"gpt-4o\",\"usage\":{\"prompt_tokens\":7,\"completion_tokens\":3}}\n\n",
            )),
            Ok(Bytes::from("data: [DONE]\n\n")),
        ];
        let s = stream::iter(chunks);
        let synth = drain_openai_stream(Box::pin(s)).await.expect("drain ok");
        assert_eq!(synth["model"], "gpt-4o");
        assert_eq!(synth["usage"]["prompt_tokens"], 7);
        assert_eq!(synth["usage"]["completion_tokens"], 3);
    }

    #[tokio::test]
    async fn drain_openai_stream_propagates_transport_errors() {
        let chunks: Vec<Result<Bytes, std::io::Error>> = vec![
            Ok(Bytes::from("data: {\"model\":\"gpt-4o\"}\n\n")),
            Err(std::io::Error::new(
                std::io::ErrorKind::ConnectionReset,
                "reset",
            )),
        ];
        let s = stream::iter(chunks);
        let res = drain_openai_stream(Box::pin(s)).await;
        match res {
            Err(StreamAggregateError::Transport(msg)) => {
                assert!(msg.contains("reset"), "unexpected error: {msg}");
            }
            other => panic!("expected transport error, got {other:?}"),
        }
    }

    #[tokio::test]
    async fn drain_anthropic_stream_aggregates_message_events() {
        let chunks: Vec<Result<Bytes, std::io::Error>> = vec![
            Ok(Bytes::from(
                "event: message_start\ndata: {\"type\":\"message_start\",\"message\":{\"model\":\"claude-3\",\"usage\":{\"input_tokens\":10,\"output_tokens\":1}}}\n\n",
            )),
            Ok(Bytes::from(
                "event: message_delta\ndata: {\"type\":\"message_delta\",\"usage\":{\"output_tokens\":42}}\n\n",
            )),
            Ok(Bytes::from(
                "event: message_stop\ndata: {\"type\":\"message_stop\"}\n\n",
            )),
        ];
        let s = stream::iter(chunks);
        let synth = drain_anthropic_stream(Box::pin(s)).await.expect("drain ok");
        assert_eq!(synth["model"], "claude-3");
        assert_eq!(synth["usage"]["input_tokens"], 10);
        assert_eq!(synth["usage"]["output_tokens"], 42);
    }
}
